### PR TITLE
Add support for labels in read token

### PR DESCRIPTION
**Pull request by StevenLeRoux**

### CODE REVIEW

**Comment by StevenLeRoux** [comment]
Mentioned in issues #173

File: warp10/src/main/java/io/warp10/continuum/Tokens.java [hunk 1]
```diff
@@ -329,6 +332,10 @@ public static ReadToken extractReadToken(String token) throws WarpScriptExceptio
       }
     }
 
+    if (!labels.isEmpty()) {
```
Comment (hbs) [inline warp10/src/main/java/io/warp10/continuum/Tokens.java]: Those labels should not be set after the OWNER label has been set.

File: warp10/src/main/java/io/warp10/continuum/Tokens.java [hunk 2]
```diff
@@ -329,6 +332,10 @@ public static ReadToken extractReadToken(String token) throws WarpScriptExceptio
...
+      labelSelectors.putAll(labels);
```
Comment (hbs) [inline warp10/src/main/java/io/warp10/continuum/Tokens.java]: owner/app/producer labels should be forbidden in the ReadToken

File: token/src/main/thrift/io_warp10_quasar_token_thrift_data.thrift [hunk 3]
```diff
@@ -139,6 +139,11 @@ struct ReadToken {
     13:optional map<string,string> attributes,
 
     /**
+     * Token labels
```
Comment (hbs) [inline token/src/main/thrift/io_warp10_quasar_token_thrift_data.thrift]: These are not token labels but extra labels which will be used  for selecting GTS. The comment should reflect that.

File: warp10/src/main/java/io/warp10/continuum/Tokens.java [hunk 4]
```diff
@@ -33,11 +33,7 @@
 import java.io.FileReader;
 import java.nio.ByteBuffer;
 import java.nio.ByteOrder;
-import java.util.ArrayList;
-import java.util.HashMap;
-import java.util.List;
-import java.util.Map;
-import java.util.UUID;
+import java.util.*;
```
Comment (hbs) [inline warp10/src/main/java/io/warp10/continuum/Tokens.java]: can you remove wildcard import and replace it with the original import.

File: warp10/src/main/java/io/warp10/continuum/Tokens.java (hunk 2 above)
Comment (StevenLeRoux) [inline warp10/src/main/java/io/warp10/continuum/Tokens.java]: Right, I've pushed a new version w/ labels being read first, and preventing owner/app/producer being forged into extra labels. Btw it breaks the ability to scope a wildcard read token through its extra labels. (not sure it has any sense anyway)

File: warp10/src/main/java/io/warp10/continuum/Tokens.java [hunk 5]
```diff
@@ -253,7 +253,28 @@ public static ReadToken extractReadToken(String token) throws WarpScriptExceptio
...
     List<String> owners = new ArrayList<String>();
     List<String> producers = new ArrayList<String>();
-    
+    Map<String, String> labels = new HashMap<String, String>();
+
+    if (rtoken.getLabelsSize() > 0) {
+      labels = rtoken.getLabels();
+      if (!labels.isEmpty()) {
+        for (Map.Entry<String, String> entry : labels.entrySet())
+        {
```
Comment (hbs) [inline warp10/src/main/java/io/warp10/continuum/Tokens.java]: this is not the style used in Warp 10, put the curly brace on the same line as the for/switch statements.

File: warp10/src/main/java/io/warp10/continuum/Tokens.java [hunk 6]
```diff
@@ -253,7 +253,28 @@ public static ReadToken extractReadToken(String token) throws WarpScriptExceptio
...
+          switch (entry.getKey())
+          {
+            case Constants.OWNER_LABEL:
+              continue;
```
Comment (hbs) [inline warp10/src/main/java/io/warp10/continuum/Tokens.java]: the first two continue statements could be removed with a fall through to the PRODUCER_LABEL case.

File: warp10/src/main/java/io/warp10/continuum/Tokens.java (hunk 5 above)
Comment (StevenLeRoux) [inline warp10/src/main/java/io/warp10/continuum/Tokens.java]: fixed

File: warp10/src/main/java/io/warp10/continuum/Tokens.java (hunk 6 above)
Comment (StevenLeRoux) [inline warp10/src/main/java/io/warp10/continuum/Tokens.java]: done

**Comment by d33d33** [comment]
👍

**Comment by d33d33** [comment]
Any news? Really need this one :)

File: warp10/src/main/java/io/warp10/continuum/Tokens.java [hunk 7]
```diff
@@ -253,7 +253,27 @@ public static ReadToken extractReadToken(String token) throws WarpScriptExceptio
     
     List<String> owners = new ArrayList<String>();
     List<String> producers = new ArrayList<String>();
-    
+    Map<String, String> labels = new HashMap<String, String>();
+
+    if (rtoken.getLabelsSize() > 0) {
+      labels = rtoken.getLabels();
+      if (!labels.isEmpty())
```
Comment (hbs) [inline warp10/src/main/java/io/warp10/continuum/Tokens.java]: Can you change the style for this file, opening braces are to appear on the same line as the for/if

File: token/src/main/java/io/warp10/quasar/encoder/QuasarTokenEncoder.java [hunk 8]
```diff
@@ -89,6 +99,12 @@ public ReadToken getReadToken(String appName, String producerUID, java.util.List
       token.setProducers(new ArrayList<ByteBuffer>());
     }
 
+
+    // hooks
```
Comment (slambour) [inline token/src/main/java/io/warp10/quasar/encoder/QuasarTokenEncoder.java]: comment not appropriate (copy / paste)

File: warp10/src/main/java/io/warp10/continuum/Tokens.java [hunk 9]
```diff
@@ -286,7 +306,9 @@ public static ReadToken extractReadToken(String token) throws WarpScriptExceptio
         labelSelectors.put(Constants.APPLICATION_LABEL, sb.toString());        
       }
     }
-    
+
+
```
Comment (slambour) [inline warp10/src/main/java/io/warp10/continuum/Tokens.java]: Removes unnecessary empty lines

**Comment by hbs** [comment]
Thanks for the PR.

**Pull request merged**